Remove dependency on numpy for serialization for XLA/open registration devices without numpy

mikaylagawarecki opened the pull request:
Related: https://github.com/pytorch/xla/issues/7799#issuecomment-2375818263 

Follow ups: Do the same for maia and mtia

## Motivation

With the move to `weights_only` by default, we are making an explicit decision not to allowlist GLOBALs required to deserialize `numpy` tensors  by default. The implication is that backends relying on numpy for serialization will fail loudly when `torch.load` flips `weights_only`.

However, we make the observation that this dependency on numpy was legacy and is not actually needed anymore. So we can remove it, which aligns with our weights_only strategy.

## Why is this ok?

The following comment on why numpy is necessary for serialization is legacy

https://github.com/pytorch/pytorch/blob/c87c9f0a01f4840bd19ac5058960c9766dd15ef8/torch/_tensor.py#L303-L312

We no longer do the following, though it was the case 5 years ago in the PR that added this 
> CPU storage is reconstructed with randomly initialized data, moved onto backend device, and then storage is updated to the serialized content

**Instead what now happens is that CPU storage is constructed with data from the file **and then** moved onto backend device.**


Old behavior (`legacy_load`): https://github.com/ailzhang/pytorch/blob/67adda891a839691790a0dcd99062430050eff3b/torch/serialization.py#L620



Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137602
* #137600
* __->__ #137444

mikaylagawarecki committed:
Remove dependency on numpy for serialization for XLA/open registration devices without numpy

[ghstack-poisoned]

albanD commented on test/test_cpp_extensions_open_device_registration.py:
```diff
@@ -529,35 +530,110 @@ def test_open_device_tensorlist_type_fallback(self):
         default_protocol = torch.serialization.DEFAULT_PROTOCOL
-        # This is a hack to test serialization through numpy
+
+        # Legacy data saved with _rebuild_device_tensor_from_numpy via
```
nit add at which version in the past this was generated with, I guess 2.4.1 ?

albanD reviewed (approved):
Sounds good!

mikaylagawarecki committed:
Update on "Remove dependency on numpy for serialization for XLA/open registration devices without numpy"

Related: https://github.com/pytorch/xla/issues/7799#issuecomment-2375818263 

Follow ups: Do the same for maia and mtia

## Motivation

With the move to `weights_only` by default, we are making an explicit decision not to allowlist GLOBALs required to deserialize `numpy` tensors  by default. The implication is that backends relying on numpy for serialization will fail loudly when `torch.load` flips `weights_only`.

However, we make the observation that this dependency on numpy was legacy and is not actually needed anymore. So we can remove it, which aligns with our weights_only strategy.

## Why is this ok?

The following comment on why numpy is necessary for serialization is legacy

https://github.com/pytorch/pytorch/blob/c87c9f0a01f4840bd19ac5058960c9766dd15ef8/torch/_tensor.py#L303-L312

We no longer do the following, though it was the case 5 years ago in the PR that added this 
> CPU storage is reconstructed with randomly initialized data, moved onto backend device, and then storage is updated to the serialized content

**Instead what now happens is that CPU storage is constructed with data from the file **and then** moved onto backend device.**


Old behavior (`legacy_load`): https://github.com/ailzhang/pytorch/blob/67adda891a839691790a0dcd99062430050eff3b/torch/serialization.py#L620






[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge
